src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java

```diff
@@ -0,0 +1,140 @@
+package com.enderio.base.common.loot.providers.nbt;
+
+import com.enderio.base.common.init.EIOLootNbtProvider;
+import com.google.common.collect.ImmutableSet;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import java.util.Set;
+import javax.annotation.Nullable;
+import net.minecraft.advancements.critereon.NbtPredicate;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.storage.loot.GsonAdapterFactory;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
+import net.minecraft.world.level.storage.loot.providers.nbt.LootNbtProviderType;
+import net.minecraft.world.level.storage.loot.providers.nbt.NbtProvider;
+
+/**
+ * A NbtProvider that provides either the
+ * {@linkplain LootContextParams#BLOCK_ENTITY block entity}'s NBT data or an
+ * entity's NBT data based on an {@link LootContext.EntityTarget} but without the id and positional data.
+ */
+public class ImprovedContextNbtProvider implements NbtProvider {
```

Comment:
What is the motive behind needing this over the old system?

Reply:
As with the other checks above the reason for this is that when the `BlockEntity` is converted over into an Item after being broken, it retains extra unneeded `BlockEntity` information in the NBT data. So, for example, when a Fluid Tank is spawned/crafted into the inventory, its item NBT data is `{id: "enderio:fluid_tank", Count: 1b}`. When it is placed into the world, then immediately broken again, its item NBT data (without this class) becomes: `{id: "enderio:fluid_tank", Count: 1b, tag: {BlockEntityTag: {x: ..., y: ..., z: ..., id: "enderio:fluid_tank"}}}`. Again, this has the annoying issue of the Fluid Tank item not stacking with another empty Fluid Tank.

This issue comes from the `ContextNbtProvider` calling `saveWithFullMetadata` on the `BlockEntity`, writing the `x`, `y`, `z`, and `id` to the NBT Tag. This provider is a carbon copy of the `ContextNbtProvider` class, except it calls `saveWithoutMetadata` on the `BlockEntity`, and returns `null` if that returns `null`. I would have just extended the `ContextNbtProvider` class, but the relevant methods to override are `private`.

As I mentioned above, the other exclusion of tags being written could be performed here as well.